agent/tool-scripts/README.md

```diff
@@ -0,0 +1,221 @@
+
+The scripts in this directory represent the pbench-agent "tools" that users
+can register to collect data from the systems related to a workload.  Each
+tool available for use with a `pbench-register-tool --name=<tool>` CLI
+invocation is listed in the `meta.json` file as either "transient" or
+"persistent".
+
+The Tool Meister running on each host where tools are registered is
+responsible for invoking that tool script to collect its generated data.  The
+commands to start and stop tools are sent to each Tool Meister at the
+appropriate time via the `pbench-init/start/stop/end/send/postprocess-tools`
+CLI interfaces.
+
+The tool scripts in this directory provide the interfaces used by the Tool
+Meister for checking for prerequisites, starting the tool data capture
+("datalog" scripts), stopping the tool data capture, and optionally
+post-processing the captured tool data ("postprocess" scripts).
+
+Most of the pbench-agent "tools" are symlinks to one script named `base-tool`.
+This allows us to share all the common behaviors between the tools, while
+handling differences where appropriate.
+
+The `external-data-source`, `node-exporter`, and `dcgm` "tools" are handled
+separately since they operate a bit differently from the rest.  See below for
+more details.
+
+## The Tool Life Cycle
+
+There are 4 phases that the Tool Meister sub-system offers. The first phase is
+for executing all the "persistent" tools.  Once the Tool Meister sub-system is
+up and running, this phase is started first (required), and is ended before
+the Tool Meister sub-system is shut down (required).  The `pbench-init-tools`
+CLI interface begins the "persistent" tools phase, and the `pbench-end-tools`
+closes out the phase.  More about "persistent" tools below.
+
+The second phase is for executing all the "transient" tools (described below).
+This phase must be started after the "persistent" tools phase begins, and must
+end before the "persistent" tools phase ends.  However, the phase can be
+repeated as many times as the caller likes.  The `pbench-start-tools` CLI
+begins the phase by starting all the "transient" tools.  Then the caller must
+invoke the `pbench-stop-tools` CLI to shut down all the "transient" tools,
+ending this phase.
+
+The third phase is for all transient tools to send their data back to the host
+driving all the local and remote Tool Meisters.  This phase can happen after
+all "transient" tool phases have executed, but must occur before ending the
+"persistent" tools phase.  The execution of the `pbench-send-tools` CLI
+interface encapsulates this phase.
+
+The fourth, and final, phase is the post-processing for "transient" tool data.
+All the post-processing for a given "transient" tool phase is invoked locally
+on the host driving all the Tool Meisters.  The execution of the
+`pbench-postprocess-tools` CLI interface handles this phase for a given
+"transient" tool phase, and must be invoked after the `pbench-send-tools` for
+a given phase.
```

Comment:
If you see the commit message for 23cc097e2857e8069c3a4c6ab485a27ed6c19d73, when PR #1248 landed, the post-processing steps only execute locally.  So once all the collected tool data is sent, then it can be post-processed.  Does that make sense?

Reply:
We can track this answer further in PR #1980.